Otus spring dev. testingapp. homework1

geracimov opened the pull request:
добавлен интерфейс TestDataService для получения данных о тестах, имплементация LocalFileTestDataService загружает данные из локальных файлов.

добавлен интерфейс TestService для тестирования по данным из TestDataService. Имплементация SimpleTestService считает баллы по количеству угаданных правильных ответов, без учета неправильных.

добавлены тесты

Commit:
добавлен интерфейс TestDataService для получения данных о тестах, имплементация LocalFileTestDataService загружает данные из локальных файлов.
добавлен интерфейс TestService для тестирования по данным из TestDataService. Имплементация SimpleTestService считает баллы по количеству угаданных правильных ответов, без учета неправильных.
добавлены тесты

saroff commented on src/main/java/ru/otus/hw1/TestingApp.java:
```diff
@@ -0,0 +1,70 @@
+
+public class TestingApp {
+
+    public static void main(String[] args) {
```
Так много логики в main классе не хорошо. Хорошая практика держать main класс всегда минималистичным.

saroff commented on src/main/java/ru/otus/hw1/TestingApp.java:
```diff
@@ -0,0 +1,70 @@
+
+        boolean repeatTest = true;
+        while (repeatTest) {
+            System.out.println("Available tests:\n" + service.getAvailTests());
```
Все общение с пользователем хорошо бы вынести в отдельный сервис, как раз main класс освободится.

saroff commented on src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java:
```diff
@@ -0,0 +1,104 @@
+    @Override
+    public Test getTest(String testName) {
+        if (testName == null
+            || "".equals(testName)
```
Есть метод String.isEmpty :) ну или можно размер строки проверить.

geracimov replied:
```suggestion
            || !testName.isEmpty()
```
Спасибо за подсказку, забыл совсем про этот метод. Наверно можно было бы даже использовать `isBlank()`, но наверно ничто не мешает завести файл с именем `     .csv`.
Заменил метода на `isEmpty()`

saroff commented on src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java:
```diff
@@ -0,0 +1,104 @@
+    public Test getTest(String testName) {
+        if (testName == null
+            || "".equals(testName)
+            || tests.get(testName) == null) {
```
А мапа нам уже null вернет, если такого ключа нету

geracimov replied:
```suggestion
```

Да, действительно, не подумал об этом. 
Убрал лишний вызов.

saroff commented on src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java:
```diff
@@ -0,0 +1,104 @@
+        Path path = tests.get(testName);
+
+        try {
+            List<Question> questions = Files.lines(path)
```
Мы сначала сканируем файлы, а потом читаем по одному, интересно :) Правда возможно стоит подумать о том, чтобы читать файлы заранее. В файловую систему лучше лишний раз не ходить.

saroff commented on src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java:
```diff
@@ -0,0 +1,104 @@
+        return new Question(description[0], choices);
+    }
+
+    private List<Choice> buildChoices(String[] stringChoises) {
```
Получается мы в двух местах обрабатываем формат массива чтобы пропустить первый элемент. допольно хрупко, потому что в случае если мы изменим формат, нужно будет не забыть поменять его везде. лучше в одном месте разбирать массив и в остальные методы передавать уже "разобранный".

saroff commented on src/main/java/ru/otus/hw1/domain/Answer.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+@Data
```
Ломбок! Супер :) С ним код гораздо приятнее.

saroff commented on src/main/java/ru/otus/hw1/service/SimpleTestService.java:
```diff
@@ -0,0 +1,122 @@
+            return null;
+        }
+        return this.test.getQuestions()
+                        .get(answers.size());
```
Неочевидно получается :) Лучше отдельно держать индекс текущего вопроса.

geracimov replied:
вынес в отдельное поле класса

geracimov committed:
Update src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java

geracimov committed:
Update src/main/java/ru/otus/hw1/dao/LocalFileTestDataService.java

Commit:
Устранение замечаний в методе getTest(testName)

Commit:
вынес индекс текущего вопроса в отделое поле класса SimpleTestService

Commit:
Вынес логику общения с пользователем в интерфейс UserCommunicator.
Добавил имплементацию ConsoleUserCommunicator, добавил бин и добавил зависимость от бина simpleTestService

Commit:
пофиксил ошибку, в двух местах был объявлен различный признак правильного варианта ответа

Commit:
- добавил импорт всех тестов на этапе инициализации, без дальнейшего чтения с файловой системы.
- добавил проверку на существование введенного пользователем наименования теста
